optuna/storages/_rdb/storage.py

```diff
@@ -39,6 +41,30 @@
 _logger = optuna.logging.get_logger(__name__)
 
 
+@contextmanager
+def _scoped_session(scoped_session: orm.scoped_session) -> Generator[orm.Session, None, None]:
+    session = scoped_session()
+    try:
+        yield session
+        session.commit()
+    except IntegrityError as e:
+        _logger.debug(
+            "Ignoring {}. This happens due to a timing issue among threads/processes/nodes. "
+            "Another one might have committed a record with the same key(s).".format(repr(e))
+        )
+        session.rollback()
+        raise
+    except SQLAlchemyError as e:
+        session.rollback()
+        message = (
+            "An exception is raised during the commit. "
+            "This typically happens due to invalid data in the commit, "
+            "e.g. exceeding max length. "
+            "(The actual exception is as follows: {})".format(repr(e))
+        )
+        raise optuna.exceptions.StorageInternalError(message).with_traceback(sys.exc_info()[2])
```

Comment:
How about explicitly catching and handling all other types of exceptions? I'm not so sure whether the indentations are correct, though.
```suggestion
        raise optuna.exceptions.StorageInternalError(message).with_traceback(sys.exc_info()[2])
    except:
        session.rollback()
        raise
    finally:
        session.close()
```

Reply:
Instead of close, I used remove. https://docs.sqlalchemy.org/en/13/orm/contextual.html#using-thread-local-scope-with-web-applications

Reply:
When I add finally section, the tests fail. I will check the reason.

Reply:
I found the reason. If we add `close` here, the code try to access the DB object after closing session.

https://github.com/optuna/optuna/pull/1628/files#diff-6048e6b64855f2df6bcd963dee2c988e671a2971a5aad48ef5e88c5f6d49b3adL1262-L1272

The current code is not compatible with `close`, but if I will modify everything, I can add `close` here.
I think it should be done in another PR.

Reply:
Sorry I don't quite understand what you mean by modifying everything but do you mind explaining what sorts of changes would be required? Does it fail when trying to close after a rollback or commit?

Reply:
In the current implementation, I enclosed code from `session = self.scoped_session()` to `self._commit(session)`  by `with` statement. But some lines access db values after `self._commit(session)` e.g. https://github.com/optuna/optuna/pull/1628/files#diff-6048e6b64855f2df6bcd963dee2c988e671a2971a5aad48ef5e88c5f6d49b3adR274 .
If I add `session.close()` here, we will have errors.

After face2face discussion with you, we concluded that we will solve this problem in another PR.